### ISSUE TITLE
Metadata construction / put-mapping requests takes too long

### ISSUE THREAD

**Issue by jonaf** [elastic/elasticsearch#38654]
**Describe the feature**: ES should not take so long to update mappings when indexing with many aliases and dynamic mappings.

In #12202, @martijnvg made some changes/improvements to MetaData, constructing a SortedMap of indices and aliases when new instances of MetaData are built.

Recently, I began the work to upgrade from our current version of Elasticsearch (1.7.3) to the last release of 2.x (2.4.6). In doing so, I have found that our use-case is no longer satisfied by Elasticsearch as a result of this particular change. And before you stop me and say that 2.x is no longer being worked on, I would point out that these changes are present and mostly unchanged in the 5.x, 6.x and master branches, so it's my anticipation that the problem I'm experiencing is persistent across all available ES releases since 2.x, I'm simply "starting from" 2.4.6.

The symptom: during heavy bulk indexing operations, I am seeing the elected master spending many, many minutes updating indices' mappings.

The cause: two-fold.

One, the mappings for my indices includes a dynamic template, which I use to disable indexing for all newly discovered fields. Why do this instead of simply set dynamic: false or strict? Because I *do* need the dynamic template feature for a specific set of fields matching a pattern, but I do *not* need to index other fields. There's currently no setting or means by which I can disable indexing fields by default without using a match-all dynamic template to do so *and* enable dynamic mapping of certain fields. As a test, I did disable the match-all dynamic template, which reduced the number of `ProcessClusterEventTimeoutException`s wildly, but still the indexing performance of the cluster was hamstrung by the elected master's updating of cluster state.

Two, in order to maximize performance (and yes, a certain level of indexing performance is *required*), I am furthermore "packing" multiple logical data sets into each index in order to reduce the number of indices and shards in the cluster. I have dwindled this number down to the ~300 range with no improvement (in ES 1.7, my cluster has ~5.5k shards, no problem). Of course, as part of this packing capability, I have some ~575k aliases pointing at the various indices and applying per-logical data set filters. I reduced the number of aliases dramatically, down to ~150k, by eliminating any unnecessary alias creation, leaving only the minimum (about equal to the number of logical data sets). This helped performance significantly, as I was no longer getting exeptions (`ProcessClusterEventTimeoutException[failed to process cluster event (put-mapping [entity]) within 30s]`), but the elected master was still doing more CPU than the ES data nodes.

I used the hot threads API to sample the ES master node during heavy indexing operations, which is what lead me to discovering #12202 as the root cause. I have added a sample to the bottom of this issue (it's long, so I posted it at the end). The core of the issue is the cost of updating and traversing a TreeMap is significantly more expensive than the previous HashMap implementations; and, moreover, ES2 seems to broadcast cluster state updates at much higher frequency, exacerbating the effects of the slower data struture. 

**What am I asking for?** Firstly, I am completely willing and eager to remediate this issue and submit a pull request for your review, in hopes that the issue can be resolved and included in future versions of Elasticsearch. In other words, I have no expectation that someone will just fix this for me. :) However, having taken a good hard look at the extent of the ES codebase, I have to admit that there's a significant lack of context on my part, which means that my naive adaptations to the code to fix the issue will likely not be up to par with Elastic's standard of quality or application design. Thus, I'm hopeful for some guidance on the ideal approach to implement a resolution to this issue. I have taken a look at the original issue and a couple related issues (mainly #12058), and it's evident that @martijnvg had some ideas about where to go next, but likely the priority of future work in this area has been deprioritized. I'm hopeful to pick it up. I do have some specific questions about it.

- I looked at just replacing the TreeMap with a SkipListMap, but that's not really sufficient to address the problem.
- As a prototype, I did a quick and dirty implementation that uses a static `ConcurrentSkipListMap` for the alias and index metadata within `IndexNameExpressionResolver`, which is diff'd and updated each time a new instance of MetaData is constructed via `MetaData#build()`. This mostly works, but there are some danger zones and as you can imagine, a lot of potential issues with concurrency, so I'm dubious about pursuing this approach with much depth.
- There are a lot of comments about moving data structures / various bits out of `MetaData` into `IndexNameExpressionResolver`. I'm not clear based on these comments what design the authors had in mind, and I'm hopeful that the context can be shared. Given my total lack of experience with the codebase, I'm sort of fishing here for a little guidance, basically. I wonder if the intention was to convert `IndexNameExpressionResolver` to a singleton or similar, or if some other design was intended.
- I haven't attempted merely implementing a diff within `MetaData#build()`. I didn't see how this could really work without changes larger in scope.

```
# curl localhost:9200/_nodes/_local/hot_threads
::: {es-host-188-161}{4Rrhm_MXTpeX4O8d-LUTOA}{IP_REDACTED}{IP_REDACTED:9300}{availability_zone=us-east-1b, client=false, data=false, master=true}
   Hot threads at 2019-01-11T22:47:16.756Z, interval=500ms, busiestThreads=3, ignoreIdleThreads=true:

   100.7% (503.5ms out of 500ms) cpu usage by thread 'elasticsearch[es-host-188-161][clusterService#updateTask][T#1]'
     4/10 snapshots sharing following 13 elements
       java.util.TreeMap.getEntry(TreeMap.java:359)
       java.util.TreeMap.get(TreeMap.java:278)
       org.elasticsearch.cluster.metadata.MetaData$Builder.build(MetaData.java:1078)
       org.elasticsearch.cluster.ClusterState$Builder.metaData(ClusterState.java:594)
       org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.applyRequest(MetaDataMappingService.java:341)
       org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.execute(MetaDataMappingService.java:230)
       org.elasticsearch.cluster.service.InternalClusterService.runTasksForExecutor(InternalClusterService.java:480)
       org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:784)
       org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:231)
       org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:194)
       java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
       java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
       java.lang.Thread.run(Thread.java:748)
     3/10 snapshots sharing following 11 elements
       org.elasticsearch.cluster.metadata.MetaData$Builder.build(MetaData.java:1081)
       org.elasticsearch.cluster.ClusterState$Builder.metaData(ClusterState.java:594)
       org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.applyRequest(MetaDataMappingService.java:341)
       org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.execute(MetaDataMappingService.java:230)
       org.elasticsearch.cluster.service.InternalClusterService.runTasksForExecutor(InternalClusterService.java:480)
       org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:784)
       org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:231)
       org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:194)
       java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
       java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
       java.lang.Thread.run(Thread.java:748)
     3/10 snapshots sharing following 11 elements
       org.elasticsearch.cluster.metadata.MetaData$Builder.build(MetaData.java:1091)
       org.elasticsearch.cluster.ClusterState$Builder.metaData(ClusterState.java:594)
       org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.applyRequest(MetaDataMappingService.java:341)
       org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.execute(MetaDataMappingService.java:230)
       org.elasticsearch.cluster.service.InternalClusterService.runTasksForExecutor(InternalClusterService.java:480)
       org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:784)
       org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:231)
       org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:194)
       java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
       java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
       java.lang.Thread.run(Thread.java:748)
```

**Comment by jonaf** [comment]
How can I draw attention to this? I have found a way to resolve the issue for my needs, but ideally the kind of issue needs to be addressed holistically, taking into account the overall design and goals.

**Comment by mayya-sharipova** [comment]
@jonaf Sorry that it took long time to reply. We indeed,  may benefit from some refactoring in Metadata.  For now, I would like to understand what template do you use to disable indexing fields (can you provide your commands here)?  I wonder if disabling doesn't actually work in your case and it leads to a lot of changes in the mappings?

**Comment by jonaf** [comment]
Hi @mayya-sharipova , thank you for your reply. I'm happy to open a PR against this issue to show how I solved it for my use-case, but it only remediates the issue fully for a very specific case (slowness due to dynamic mapping updates during indexing) and does not help in the general case. (Note that my PR is against 2.4.6, but the changes are pretty easy to adapt for later versions of ES, and of course I'm happy to do that as well.)

To answer your question, the dynamic template I use is much too large to paste here -- we use Elasticsearch at Bazaarvoice to index over 1.2 billion documents (all of which are queryable at anytime), and many of these documents contain multilingual fields, which are indexed with various analyzers. Moreover, we have many fields that are mapped dynamically (unknown at index creation time). I did run a lot of tests because I originally thought my problem was due to spurious mapping updates, but after some experimentation I found that Elasticsearch is doing a great job of only invoking mapping updates when there are effective changes to documents.

I did attempt to disable dynamic mappings for sub-fields, but that had no noticeable effect on the number of updates or the slowness in general.

**Comment by DaveCTurner** [comment]
Hi @jonaf, to be clear, you are interested in the performance of the following line that still exists in `master` today?

https://github.com/elastic/elasticsearch/blob/74ba9cedcb72b5b94836e91af1a22c596b1db29c/server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java#L1193

Specifically, you are saying that this is slow because your cluster has some 500k+ aliases and this means that each call to this method takes 30+ seconds to complete? I think we'd generally consider a setup with hundreds of thousands of aliases to be an error, and would encourage you to find a different way to achieve what you're trying to achieve with this setup. That said, it's always worth contemplating this sort of improvement, as long as we can balance the complexity it introduces against the benefits.

I can't speak about the situation back in 2.4.6 but I can say that in the active development branches objects like `MetaData` are generally immutable, which hopefully means you do not need to worry deeply about concurrency. I would recommend doing your work with an actively developed version, ideally `master`, as forward-ports are tricky and any performance measurements you make against such an old version will not carry forward.

The codebase does include a microbenchmarking framework that should be suitable for this sort of work. There's a README at https://github.com/elastic/elasticsearch/tree/master/benchmarks that should help to get you started.

**Comment by jonaf** [comment]
@DaveCTurner thanks for your reply. Having spent plenty of time with 2.4.6 codebase, I can say that it was true back then that `MetaData` is treated as immutable. The issues with concurrency actually *arise* from this assumption, though -- one way to optimize the performance is to re-use the `MetaData` object instead of creating new instances all the time, because the slowness occurs when a new `MetaData` instance is created via `MetaDataBuilder#build()`.

I am completely aware that having hundreds of thousands of aliases is considered a degenerate use-case. Unfortunately, there's really no alternative for Bazaarvoice, which has been relying on this capability since 2012. It's either hundreds of thousands of indices, or hundreds of thousands of aliases, and the former has even more insurmountable problems due to the sheer magnitude of cluster state, with the additional drawback that the ability to alias an index for swapping to new indices with no downtime is a critical business use-case. I wish our situation were different, but I see no way to make it so.

I've looked at the codebase for versions 5.x, 6.x and master, and they all include code that traverses and reads + writes the `TreeMap` data structure when the `MetaData` object is built. In fact, the line you referenced has no performance impact because it returns a *view* of the underlying map, so no actual copying is happening here, just the methods that mutate the map are shadowed to prevent modification to the map.

The offending line, if I had to choose one, is https://github.com/elastic/elasticsearch/blob/74ba9cedcb72b5b94836e91af1a22c596b1db29c/server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java#L1218 in current master. But really that whole method, and the structure of it, is extremely complex, I think it's `O(n * 2 log m)` on a TreeMap (so each operation itself is of complexity `O(log n)`. I have an implementation (in 2.4.6) that improves it by performing the work in `O(n)` time in the special case where the list of aliases and indices hasn't changed since the last `MetaData` object was initialized. In my performance tests, the results have been so dramatic, it out-performed Elasticsearch 1.7.3 by a wide margin (about twice).

Sadly, my fix works most effectively when the aliases or indices does not change frequently *during* indexing, so a more general solution to the problem is likely warranted, but probably a lot more involved.

Since I don't expect to get much help fixing this simply by opening an issue, here's how I'd like to proceed. Please advise.

1. Submit my pull request against master, including benchmarking. I hope this will show you guys that I'm serious about making changes. :) 
2. Elastic can review it, hopefully accept it, but at the very least, understand the very serious problem the use of `TreeMap` introduced in ES 2.0.0-alpha1, but more generally, the problem with the current design of rebuilding the `aliasAndIndexLookup` data structure so frequently and unnecessarily.
3. If the PR is denied, I'm happy to continue working with Elastic on a change that is broader (i.e., design-level changes to avoid the problem of rebuilding the `aliasAndIndexLookup` data structure within `MetaData` objects -- the ideal solution, I think).

Does that sound good?

**Comment by DaveCTurner** [comment]
> The offending line, if I had to choose one, is

Yes, that line is within the `buildAliasAndIndexLookup()` method called on the line I quoted. Good, we're talking about the same thing.

>  my fix works most effectively when the aliases or indices does not change frequently during indexing, so a more general solution to the problem is likely warranted, but probably a lot more involved.

Leaving the aliases/indices unchanged is a common case, so a simple fix that improves performance in only that case could well be preferable to a more involved change. As I said, we need to balance the complexity of any change against its benefits, and the benefits here might only be felt by a tiny number of users that require hundreds of thousands of aliases.

Please go ahead and open a PR so we can see more clearly what change you are proposing and how involved it is. Benchmarks and tests would be nice, and will be essential before any change is merged, but I also have some concerns about the design that might be worth discussing before you put a lot of effort into validating it.

**Comment by jonaf** [comment]
OK, thanks, @DaveCTurner ! I'll open a pull request against master in short order. I appreciate your attention to the issue!

**Comment by jonaf** [comment]
@DaveCTurner I've opened PR #40432 for this issue against master. There is one additional optimization I did not include, in the interest of getting the more impactful (but smaller) change resolved, but I'm happy to make that change if you're interested.

I had trouble with `./gradlew check` failing after running for 1-2 hours, but it fails for a variety of cryptic reasons. This is probably due to my lack of familiarity (running the tests in 2.x using Maven was easier for me, but I'm very comfortable with Maven). I'll continue trying to get the tests to complete, but in the meantime, at least we can discuss/review the PR itself and talk about the best benchmarks to perform against it. (I did look at Rally, but it looks like Rally might need some additional functionality added to its core to produce the kind of slowness I experienced in my degenerate use-case.)

**Comment by jonaf** [comment]
During deeper investigation on later versions of Elasticsearch (5.x and beyond) in PR #40432 , it was discovered that there is a change in 5.x (specifically, part of #21306) that resolves the problem that's happening in ES 2.4. We concluded that no changes were necessary in any version of ES >= 5.x, and so the PR has been closed, and I'm closing this issue as well.
